src/Util/PaginationHelper.php

```diff
@@ -0,0 +1,238 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Util;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use Exception;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class PaginationHelper
+{
+    private $requestStack;
+    private $resourceMetadataFactory;
+    private $resourceMetadata;
+
+    private $enabled = true;
+    private $clientEnabled = false;
+    private $clientItemsPerPage = false;
+    private $itemsPerPage = 30;
+    private $maximumItemsPerPage = null;
+
+    private $parameterNamePage = 'page';
+    private $parameterNameEnabled = 'pagination';
+    private $parameterNameItemsPerPage = 'itemsPerPage';
+
+    private $resourceClass;
```

Comment:
A service must not have mutable states like that. This class shouldn't be a service, it should be instantiated directly in the extension, it will allow to get rid of those states, and it's a cleaner design (better, it would be easier to be backward compatible).